polkadot/node/network/collator-protocol/src/validator_side/claim_queue_state.rs

```diff
@@ -152,904 +199,2584 @@ impl ClaimQueueState {
+		candidate_hash: CandidateHash,
+	) -> bool {
+		gum::trace!(
+			target: LOG_TARGET,
+			?para_id,
+			?relay_parent,
+			?candidate_hash,
+			"second_at"
+		);
 
-		window.filter(|b| !b.claimed).filter_map(|b| b.claim).collect()
-	}
-}
+		if self.candidates.get(relay_parent).map_or(false, |c| c.contains(&candidate_hash)) {
+			let window = self.get_window_mut(relay_parent);
 
-#[cfg(test)]
-mod test {
-	use super::*;
+			for w in window {
+				if w.claimed == ClaimState::Pending(Some(candidate_hash)) ||
+					w.claimed == ClaimState::Seconded(candidate_hash)
+				{
+					w.claimed = ClaimState::Seconded(candidate_hash);
+					return true;
+				}
+			}
 
-	#[test]
-	fn sane_initial_state() {
-		let mut state = ClaimQueueState::new();
-		let relay_parent = Hash::from_low_u64_be(1);
-		let para_id = ParaId::new(1);
+			gum::warn!(
+				target: LOG_TARGET,
+				?para_id,
+				?relay_parent,
+				?candidate_hash,
+				"Hash found in candidates but can't find a claim for it. This should never happen"
+			);
 
-		assert!(!state.can_claim_at(&relay_parent, &para_id));
-		assert!(!state.claim_at(&relay_parent, &para_id));
-		assert_eq!(state.unclaimed_at(&relay_parent), vec![]);
+			return false
+		} else {
+			// this is a new claim
+			self.find_a_free_claim(
+				relay_parent,
+				para_id,
+				Some(candidate_hash),
+				ClaimState::Seconded(candidate_hash),
+			)
+		}
 	}
 
-	#[test]
-	fn add_leaf_works() {
-		let mut state = ClaimQueueState::new();
-		let relay_parent_a = Hash::from_low_u64_be(1);
-		let para_id = ParaId::new(1);
-		let claim_queue = vec![para_id, para_id, para_id];
-
-		state.add_leaf(&relay_parent_a, &claim_queue);
-		assert_eq!(state.unclaimed_at(&relay_parent_a), vec![para_id, para_id, para_id]);
-
-		assert_eq!(
-			state.block_state,
-			VecDeque::from(vec![ClaimInfo {
-				hash: Some(relay_parent_a),
-				claim: Some(para_id),
-				claim_queue_len: 3,
-				claimed: false,
-			},])
-		);
-		assert_eq!(
-			state.future_blocks,
-			VecDeque::from(vec![
-				ClaimInfo { hash: None, claim: Some(para_id), claim_queue_len: 1, claimed: false },
-				ClaimInfo { hash: None, claim: Some(para_id), claim_queue_len: 1, claimed: false }
-			])
+	/// Returns `true` if there is a free spot in claim queue (free claim) for `para_id` at
+	/// `relay_parent`. The function only performs a check. No actual claim is made.
+	pub(crate) fn can_claim_at(
+		&mut self,
+		relay_parent: &Hash,
+		para_id: &ParaId,
+		maybe_candidate_hash: Option<CandidateHash>,
+	) -> bool {
+		gum::trace!(
+			target: LOG_TARGET,
+			?para_id,
+			?relay_parent,
+			"can_claim_at"
 		);
 
-		// should be no op
-		state.add_leaf(&relay_parent_a, &claim_queue);
-		assert_eq!(state.block_state.len(), 1);
-		assert_eq!(state.future_blocks.len(), 2);
-
-		// add another leaf
-		let relay_parent_b = Hash::from_low_u64_be(2);
-		state.add_leaf(&relay_parent_b, &claim_queue);
+		self.find_a_free_claim(relay_parent, para_id, maybe_candidate_hash, ClaimState::Free)
```

Comment:
Here `self.find_a_free_claim()` might return `true` even if there is no claim with status `ClaimState::Free`. For example if `maybe_candidate_hash = Some(candidate_hash`) and if there is a claim with status `ClaimState::Seconded(candidate_hash)`,  for the provided `relay_parent` and `para_id`. 

Is this correct ? If so, we should probably rename `find_a_free_claim`. If not, could we remove the `maybe_candidate_hash` param from `find_a_free_claim` and only look at the `candidate_hash` in the `new_state` if present ?

Reply:
Clarified